### PR TITLE
Quickly fix HKT GADT unsoundness issue

### LINKED ISSUE #9044: Verify HKT GADT code
## Minimized code

```Scala
sealed trait Test[+F[_], +A] extends Product with Serializable

object Test {

  implicit class Syntax[F[_], A](val self: Test[F, A]) extends AnyVal {

    def fold[B](completed: F[A] => B): B = self match {
      case Completed(fa) => completed(fa)
    }
  }


  final case class Completed[F[_], A](fa: F[A]) extends Test[F, A]
}
```

## Output

Dotty 0.24.0-RC1:

```scala
[error] -- [E007] Type Mismatch Error: /Users/daniel/Development/Scala/cats-effect/core/src/main/scala/cats/effect/Test.scala:26:38 
[error] 26 |      case Completed(fa) => completed(fa)
[error]    |                                      ^^
[error]    |                 Found:    (fa : F$1[A$1])
[error]    |                 Required: F[A]
[error]    |
[error]    |                 where:    A$1 is a type in method fold with bounds <: A
[error]    |                           F$1 is a type in method fold with bounds <: F
```

Dotty master:
```scala
-- [E029] Pattern Match Exhaustivity Warning: try/dj.scala:7:43 ----------------
7 |    def fold[B](completed: F[A] => B): B = self match {
  |                                           ^^^^
  |                          match may not be exhaustive.
  |
  |                          It would fail on pattern case: Test.Completed(_)

longer explanation available when compiling with `-explain`
-- Warning: try/dj.scala:3:7 ---------------------------------------------------
3 |object Test {
  |       ^
  |       the type test for Test.Completed[?, ?] cannot be checked at runtime
2 warnings found
```

## Expectation

I think 0.24.0-RC1 got it right, current master is missing the error and producing some weird warnings.

### PR DESCRIPTION
**Pull request by AleksanderBG**
Check that the result of GADT approximation is actually a subtype of what we're approximating.

Fixes #9044.

### CODE REVIEW

**Comment by AleksanderBG** [comment]
@smarter so it turns out that the issue was with GADT member healing! Can you take a look at the code? The relevant test passes on my local machine. 

I'm just surprised that `tpd.Typed(t, tpt)` doesn't actually check that the type ascription is, well, an ascription. Let's consider after the release as to why GADT approximation (and type variable forcing, by extension) creates types that aren't actual subtypes of what we're approximating.

**Comment by smarter** [comment]
> I'm just surprised that tpd.Typed(t, tpt) doesn't actually check that the type ascription is, well, an ascription. 

TreeChecker should check that but doesn't, I have a branch which fixes that but it needs more work because it revealed situations where this isn't true currently.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 1]
```diff
@@ -3411,20 +3411,23 @@ class Typer extends Namer
...
-            )(using nestedCtx)
-          if (!nestedCtx.reporter.hasErrors) {
-            // GADT recovery successful
-            nestedCtx.typerState.commit()
-            res
-          } else {
-            // otherwise fail with the error that would have been reported without the GADT recovery
-            err.typeMismatch(tree, pt, failure)
+          val approximated = Inferencing.approximateGADT(wtp)(using nestedCtx)
+          if (tree.tpe <:< approximated) {
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:3415]: Shouldn't that check be done in the nestedCtx?

Comment (AleksanderBG) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala:3415]: That's sadly not the part that's making it explode. Not even making the check frozen and doing it in nestedCtx helps. I have no idea why, seems that we will need to take a more scorched-earth type of approach.

**Pull request closed without merging**